### PR TITLE
FEATURE: add semantic search with hyde bot

### PR DESCRIPTION
**Pull request by SamSaffron**
- FEATURE: add semantic search to bot
- FIX: embedding backfill rake task not working

### CODE REVIEW

File: lib/modules/ai_bot/commands/search_command.rb [hunk 1]
```diff
@@ -120,12 +124,35 @@ def process(**search_args)
         )
 
       # let's be frugal with tokens, 50 results is too much and stuff gets cut off
-      limit ||= 20
-      limit = 20 if limit > 20
+      limit ||= MAX_RESULTS
+      limit = MAX_RESULTS if limit > MAX_RESULTS
+
+      should_try_semantic_search = SiteSetting.ai_embeddings_semantic_search_enabled
+      should_try_semantic_search &&= (limit == MAX_RESULTS)
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/search_command.rb:131]: Oh I never knew about &&=, nice!

File: lib/modules/ai_bot/commands/search_command.rb [hunk 2]
```diff
@@ -120,12 +124,35 @@ def process(**search_args)
...
+      should_try_semantic_search &&= (search_args.keys - %i[search_query order]).length == 0
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/search_command.rb:132]: nit, could be `(search_args.keys - %i[search_query order]).zero?`

File: lib/modules/ai_bot/commands/search_command.rb [hunk 3]
```diff
@@ -120,12 +124,35 @@ def process(**search_args)
...
+      should_try_semantic_search &&= (search_args[:search_query].present?)
+
+      limit = limit - MIN_SEMANTIC_RESULTS if should_try_semantic_search
 
       posts = results&.posts || []
       posts = posts[0..limit - 1]
 
+      if should_try_semantic_search
+        semantic_search = DiscourseAi::Embeddings::SemanticSearch.new(Guardian.new())
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/search_command.rb:141]: I know `Guardian.new()` is making an anon user, but it kind of looks like a weird mistake...we should make a `Guardian.anonymous` method in core to do this. Anyway, doesn't need changing in this PR really, just a thought.

File: lib/modules/ai_bot/commands/search_command.rb [hunk 4]
```diff
@@ -120,12 +124,35 @@ def process(**search_args)
...
+        topic_ids = Set.new(posts.map(&:topic_id))
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/search_command.rb:142]: This is a nice way of getting uniq, I am guessing it's faster than doing `.uniq` on the end?

File: lib/modules/ai_bot/commands/search_command.rb [hunk 5]
```diff
@@ -120,12 +124,35 @@ def process(**search_args)
...
+
+        semantic_search
+          .search_for_topics(search_args[:search_query])
+          .each do |post|
+            next if topic_ids.include?(post.topic_id)
+
+            topic_ids << post.topic_id
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/search_command.rb:149]: I'm guessing it does, just confirming `<<` works for a `Set`?

**Review by martin-brennan — APPROVED** [review]

File: lib/modules/ai_bot/commands/search_command.rb (hunk 4 above)
Comment (SamSaffron) [reply lib/modules/ai_bot/commands/search_command.rb:142]: It is really splitting hairs cause there are just 20 elements here... but I wanted a Set ... not an Array so the `include?` is faster.

File: lib/modules/ai_bot/commands/search_command.rb (hunk 2 above)
Comment (SamSaffron) [reply lib/modules/ai_bot/commands/search_command.rb:132]: true ... I so rarely use zero ActiveSupport is so magical

**Pull request merged**